From 74866658f1234fbf81a795a93110b3a16ffd2bbe Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Wed, 16 Oct 2024 15:43:47 +0200
Subject: [PATCH 1331/1695] clk: stm32mp25: do not configure PLL1 MUX when
 M33TDCID

Do not access PLL mux when M33 is TDCID. Just check if the mux is correct.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 34f1f9cd9f684739fbab4b6ce1e372f5e3a386c1 clk: stm32mp25: do not configure PLL1 MUX when M33TDCID]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I06b2986faa35de703a76e454bed53c326cba6bcc
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/411803
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 927380a95..b689bf887 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -1903,6 +1903,15 @@ static int clk_stm32_pll_set_mux(struct clk_stm32_priv *priv __unused,
 	return 0;
 }
 
+static int clk_stm32_pll_check_mux(struct clk_stm32_priv *priv __unused,
+				   uint32_t src)
+{
+	int mux = (src & MUX_ID_MASK) >> MUX_ID_SHIFT;
+	int sel = (src & MUX_SEL_MASK) >> MUX_SEL_SHIFT;
+
+	return stm32_mux_get_parent(mux) != (size_t)sel;
+}
+
 static int clk_stm32_pll1_init(struct clk_stm32_priv *priv,
 			       int pll_idx __unused,
 			       struct stm32_pll_dt_cfg *pll_conf)
@@ -1918,7 +1927,11 @@ static int clk_stm32_pll1_init(struct clk_stm32_priv *priv,
 
 	stm32mp2_a35_ss_on_bypass();
 
-	ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
+	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8))
+		ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
+	else
+		ret = clk_stm32_pll_check_mux(priv, pll_conf->src);
+
 	if (ret != 0)
 		panic();
 
-- 
2.47.2

